docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,798 @@
+
+### Entity registration
+
+Since thin-edge doesn't enforce what each device identification level means,
+an explicit registration is required to register every entity that is going to send data or receive commands.
+For example, before a measurement can be sent from a service named `collectd` from the device `Rpi1001`,
+the entity named `Rpi1001` must be registered as a `device`, and `collectd` must be registered as a `service`.
+
+An entity can be registered with thin-edge by publishing a retained message to the entity identification topic prefix
+with the entity type and other metadata that defines that entity.
+To model the example mentioned above, an entity identification topic scheme like the following can be used:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+The `Rpi1001` device must be registered as a `device` by publishing the following retained message:
+
+```sh te2mqtt
+tedge mqtt pub -r tedge/_/Rpi1001 '{
+  "type": "device"
+}'
+```
+
+The device is registered under the default namespace indicated by `_` at the second level.
+The registration message supports additional fields like `name`, `parent`, `device-type` etc as well,
+which will be explained in detail later.
+
+Once the device is registered, the `collectd` service can be registered as a `service` type as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r tedge/_/Rpi1001/_/collectd '{
+  "type": "service"
+}'
+```
+
+Once these entities are registered, they can start sending data or receive commands.
+
+Examples:
+* Device measurement: `tedge/_/Rpi1001/_/_/mes/battery_reading`
+* Service measurement: `tedge/_/Rpi1001/_/collectd/mes/cpu_usage`
+
+**Main Device**
+
+The main device does not need any explicit registration and can be referred to using the `main` alias as device id.
+Hence, a measurement associated to the main device can be sent to following topic:
+
+```
+`tedge/_/main/_/_/mes/battery_reading`
+```
+
+But if some additional metadata needs to be added to the main device,
+they can be pushed to the `tedge/_/main` topic as done using a registration message.
+
+**Immediate Child Devices**
+
+Immediate child devices of the main device can be registered using the registration protocol described above.
+If a device is registered with `main` as the `parent` field value, or without a `parent` field,
+it is assumed to be an immediate child device.
+
+**Nested Child Devices**
+
+Nested child devices must always register explicitly by declaring their `parent`.
+
+### Automatic registration
+
+If an explicit registration is not done, thin-edge will assume the following topic level convention:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+and auto-register the entities as per their positions in the topics.
+
+For example, if the following measurement message is received without any explicit registrations,
+
+```
+tedge/device_namespace/Rpi1001/service_namespace/collectd/mes/cpu_usage
```

Comment:
The whole idea of <device-namespace> is to group the `devices` under some namespace right? Maybe we need concrete examples for <*-namespace> fragment of the `topic`.
It's maybe the same for other `namespaces` as well.

Reply:
The namespaces named `device_namespace` and `service_namespace` were concrete namespaces in this example. I guess I should have used more explicit names like `abc_device_namespace` and `xyz_service_namespace` or something like that.

Reply:
Yes, that would be better.

Reply:
You can just use `device` and `service` if you want to show an example of how to use the topic structure. Having `_namespace` in the suffix of the example is a bit redundant, as `device` is representing a namespace, and you have already highlighted that in the generic topic structure description.

Reply:
This example has been replaced by `te/abc/Rpi1001/xyz/collectd/m/cpu_usage` (not sure this is better :-))
and we adopted `device` and `service` as the default namespaces (e.g. `te/device/main/service/tedge-agent`).